File: app/src/main/java/de/xikolo/managers/AnnouncementManager.java
package de.xikolo.managers;

import de.xikolo.jobs.ListCourseAnnouncementsJob;
import de.xikolo.jobs.ListGlobalAnnouncementsJob;
import de.xikolo.jobs.UpdateAnnouncementVisitedJob;
import de.xikolo.jobs.base.JobCallback;
import de.xikolo.managers.base.BaseManager;
import de.xikolo.models.Announcement;
import io.realm.Realm;
import io.realm.RealmChangeListener;
import io.realm.RealmResults;
import io.realm.Sort;

public class AnnouncementManager extends BaseManager {

    public static final String TAG = AnnouncementManager.class.getSimpleName();

    public RealmResults listGlobalAnnouncements(Realm realm, RealmChangeListener<RealmResults<Announcement>> listener) {
        if (listener == null) {
            throw new IllegalArgumentException("RealmChangeListener should not be null for async queries.");
        }

        RealmResults<Announcement> announcementListPromise = realm
                .where(Announcement.class)
                .findAllSortedAsync("publishedAt", Sort.DESCENDING);

        announcementListPromise.addChangeListener(listener);

        return announcementListPromise;
    }

    public RealmResults listCourseAnnouncements(String courseId, Realm realm, RealmChangeListener<RealmResults<Announcement>> listener) {
        if (listener == null) {
            throw new IllegalArgumentException("RealmChangeListener should not be null for async queries.");
        }

        RealmResults<Announcement> announcementListPromise = realm
                .where(Announcement.class)
                .equalTo("courseId", courseId)
                .findAllSortedAsync("publishedAt", Sort.DESCENDING);

        announcementListPromise.addChangeListener(listener);

        return announcementListPromise;
    }

    public void requestGlobalAnnouncementList(JobCallback callback) {
        jobManager.addJobInBackground(new ListGlobalAnnouncementsJob(callback));
    }

    public void requestCourseAnnouncementList(String courseId, JobCallback callback) {
        jobManager.addJobInBackground(new ListCourseAnnouncementsJob(courseId, callback));
    }

    public void updateAnnouncementVisited(String announcementId) {
        jobManager.addJobInBackground(new UpdateAnnouncementVisitedJob(announcementId));
    }

}

File: app/src/main/java/de/xikolo/managers/AnnouncementManager.kt
package de.xikolo.managers

import de.xikolo.network.jobs.ListCourseAnnouncementsJob
import de.xikolo.network.jobs.ListGlobalAnnouncementsJob
import de.xikolo.network.jobs.UpdateAnnouncementVisitedJob
import de.xikolo.network.jobs.base.RequestJobCallback
import de.xikolo.models.Announcement
import io.realm.Realm
import io.realm.RealmChangeListener
import io.realm.RealmResults
import io.realm.Sort

class AnnouncementManager {

    companion object {
        val TAG: String = AnnouncementManager::class.java.simpleName
    }

    fun listGlobalAnnouncements(realm: Realm, listener: RealmChangeListener<RealmResults<Announcement>>?): RealmResults<*> {
        if (listener == null) {
            throw IllegalArgumentException("RealmChangeListener should not be null for async queries.")
        }

        val announcementListPromise = realm
            .where(Announcement::class.java)
            .sort("publishedAt", Sort.DESCENDING)
            .findAllAsync()

        announcementListPromise.addChangeListener(listener)

        return announcementListPromise
    }

    fun listCourseAnnouncements(courseId: String, realm: Realm, listener: RealmChangeListener<RealmResults<Announcement>>?): RealmResults<*> {
        if (listener == null) {
            throw IllegalArgumentException("RealmChangeListener should not be null for async queries.")
        }

        val announcementListPromise = realm
            .where(Announcement::class.java)
            .equalTo("courseId", courseId)
            .sort("publishedAt", Sort.DESCENDING)
            .findAllAsync()

        announcementListPromise.addChangeListener(listener)

        return announcementListPromise
    }

    fun requestGlobalAnnouncementList(callback: RequestJobCallback) {
        ListGlobalAnnouncementsJob(callback).run()
    }

    fun requestCourseAnnouncementList(courseId: String, callback: RequestJobCallback) {
        ListCourseAnnouncementsJob(courseId, callback).run()
    }

    fun updateAnnouncementVisited(announcementId: String) {
        UpdateAnnouncementVisitedJob.schedule(announcementId)
    }

}
